internal/netpoll/queue/queue.go

```diff
@@ -20,12 +20,30 @@
 
 package queue
 
-// Task is a asynchronous function.
-type Task func() error
+import "sync"
+
+type TaskFunc func([]byte) error
+
+// Task is a wrapper that contains function and its argument.
+type Task struct {
+	Run TaskFunc
+	Buf []byte
+}
+
+var taskPool = sync.Pool{New: func() interface{} { return &Task{} }}
+
+func GetTask() *Task {
+	return taskPool.Get().(*Task)
+}
+
+func PutTask(task *Task) {
+	task.Run, task.Buf = nil, nil
+	taskPool.Put(task)
+}
 
 // AsyncTaskQueue is a queue storing asynchronous tasks.
 type AsyncTaskQueue interface {
-	Enqueue(Task)
-	Dequeue() Task
+	Enqueue(*Task)
```

Comment:
I would hide taskPool inside `lockFreeQueue` in that `Enqueue(Task)` will get an obj from taskPool and `Dequeue() Task` puts the obj back to the taskPool - to simplify the caller side from calling get and put from the pool.

Reply:
In that case, the input parameter of `Enqueue()` and the returned value of `Dequeue()` would be changed to `f queue.TaskFunc, buf []byte`, which breaks encapsulation cuz currently I use `Task` struct as a wrapper for `AsyncTaskQueue` interface.

Reply:
You can keep the `AsyncTaskQueue` interface intact as I did here
https://github.com/lming/gnet/blob/3a24138f5029663aaec99a41c7ec1d593d6262ce/internal/netpoll/queue/lock_free_queue.go#L144

Reply:
But it's a bit urgly that the caller needs to check `Task`'s attribute of zero value to see if there is no task returned ...
So this is a minor suggestion. Feel free to go with your current way :)

Reply:
The `nodePool` which you introduce will cause data race cuz the lock-free queue built with atomic CAS is not like the queue protected by a mutex, a node from `Dequeue()` of a lock-free queue could be held by multiple callers who are calling `Dequeue()` at the same time, even though this node has been put into `nodePool` and get from `nodePool` in other callers of `Enqueue()`, that's why your original PR failed the tests under async mode.

Reply:
Hmm I don't see the data race issue in that PR, if it's true then it's a big problem ... Mind having a quick call to discuss? I will shoot you an email for the contact. Thanks.

Reply:
I would like to talk about that, but unfortunately, it seems that we are in different time zones and it's almost 3:00 am in China, I'm exhausted and have to go get asleep, could we have the discussion like 7~8 hours later? Please leave your email address here and I will send your email noticing you I'm ready for a chat, thanks~

BTW, you can run `go test -race` on your branch to see the data races I was talking about.

Reply:
Sure let's find another time for the call. I've sent an email to i@andypan.me with my contact. In the mean time I'll try the `-race` test. Good night!